eval/src/vespa/eval/instruction/mixed_inner_product_function.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "mixed_inner_product_function.h"
+#include <vespa/eval/eval/operation.h>
+#include <vespa/eval/eval/value.h>
+#include <cblas.h>
+
+namespace vespalib::eval {
+
+using namespace tensor_function;
+using namespace operation;
+
+namespace {
+
+template <typename LCT, typename RCT>
+double my_dot_product(const LCT * &lhs, const RCT * rhs, size_t count) {
+    double result = 0.0;
+    for (size_t i = 0; i < count; ++i) {
+        result += (*lhs++) * (*rhs++);
+    }
+    return result;
+}
+
+template <>
+double my_dot_product<double,double>(const double * &lhs, const double * rhs, size_t count) {
+    double result = cblas_ddot(count, lhs, 1, rhs, 1);
+    lhs += count;
+    return result;
+}
+
+template <>
+double my_dot_product<float,float>(const float * &lhs, const float * rhs, size_t count) {
+    double result = cblas_sdot(count, lhs, 1, rhs, 1);
+    lhs += count;
+    return result;
+}
+
+template <typename MCT, typename VCT, typename OCT>
+void my_mixed_inner_product_op(InterpretedFunction::State &state, uint64_t param_in) {
+    const auto &param = unwrap_param<MixedInnerProductParam>(param_in);
+    const auto &mixed = state.peek(1);
+    const auto &dense = state.peek(0);
+    auto m_cells = mixed.cells().typify<MCT>();
+    auto v_cells = dense.cells().typify<VCT>();
+    const auto &index = mixed.index();
+    size_t num_subspaces = index.size();
+    size_t num_output_cells = num_subspaces * param.out_subspace_size;
+    ArrayRef<OCT> out_cells = state.stash.create_uninitialized_array<OCT>(num_output_cells);
+    const MCT *m_cp = m_cells.begin();
+    OCT *out = out_cells.begin();
```

Comment:
consider re-writing to a modern for-loop over all output cells `for(OCT &cell: out_cells) {`